Performance Analysis Tool (#41)
raiden-network/scenario-player: analysis/README.rst

Scenario Player Analysis
========================

> This is an analysis tool to show performance metrics of a scenario using the logs of a scenario run.

Reviewer: ```suggestion
This is an analysis tool to show performance metrics using the logs of a scenario player run.
```

Prerequisites
"""""""""""""
- `pyenv <https://github.com/pyenv/pyenv>`_
- `virtualenvwrapper <https://github.com/virajkanwade/venvwrapper>`_

> You may also use other python virtual environments

Reviewer: ```suggestion
You may also use other python virtual environments.
```

Installation
""""""""""""

.. code-block:: bash

    pyenv shell 3.7.0
    mkvenv raiden-analysis
    pip install -r requirements.txt

How To Use
""""""""""
Either start the analysis directly or print out the help:

.. code-block:: bash

    python analysis.py $PATH_TO_SCENARIO_LOG_FILE

    python analysis.py -h

Result
""""""
There are currently two output files generated:

> - a gantt plot is rendered in a HTML file called (default: ``raiden-gantt-analysis.html``)

Reviewer: ```suggestion
- a gantt plot is rendered in an HTML file called (default: ``raiden-gantt-analysis.html``)
```

Author: That would be wrong then ;-)

Reviewer 2: Actually not. You spell it after how you'd pronounce it. And "H" is pronounced with an "a" sound in the beginning, so the spelling should actually be what Christian suggests.

Author: Sorry - almost forgot about that complexity

- a CSV file is created which contains the duration of the tasks: (default: ``raiden-gantt-analysis.csv``)

> - a JSON file is with the statistic summary calculated on base of transfers (default: ``raiden-scenario-player-analysis.json``)

Reviewer: ```suggestion
- a JSON file with the statistic summary on the basis of transfers is getting calculated (default: ``raiden-scenario-player-analysis.json``)
```

Limitations/Caveats
"""""""""""""""""""
- the duration of a task is done by looking at the next event in the list of events (sorted by time)

> - parallel task processing might not be visualized not correctly

Reviewer: ```suggestion
- parallel task processing might not be visualized correctly
```

> - the number of ticks on the y-axis can be to much when too many tasks have been emitted in the log, in that case uncomment the ``showticklabels`` line in the ``draw_gantt()`` method.

Reviewer: ```suggestion
- the number of ticks on the y-axis can be too much when too many tasks have been emitted in the log. In that case uncomment the ``showticklabels`` line within the ``draw_gantt()`` method.
```

> - log messages are grouped by task-id, resulting in request/response calls are grouped

Reviewer: ```suggestion
- log messages are grouped by task-id, resulting in request/response calls being grouped
```

How To Use
^^^^^^^^^^
Either start the analysis directly or print out the help:

.. code-block:: bash

    python analysis.py $PATH_TO_SCENARIO_LOG_FILE

    python analysis.py -h

See an example result below.

Examples
^^^^^^^^
Two different scenarios are currently available as examples and each have three generated files:

- A gantt plot is rendered in an HTML file - it contains a legend and a summary
- A CSV file is created which contains the duration of the tasks
- A JSON file with a statistic summary on the basis of transfers is getting calculated

Simple-Scenario
"""""""""""""""

- Gantt: `<example/simple-scenario/raiden-scenario-player-analysis.html>`_
- CSV: `<example/simple-scenario/raiden-scenario-player-analysis.csv>`_
- JSON-Summary: `<example/simple-scenario/raiden-scenario-player-analysis.json>`_

Mediated-Transfers-Scenario
"""""""""""""""""""""""""""

- Gantt: `<example/mediated-transfers-scenario/raiden-scenario-player-analysis.html>`_
- CSV: `<example/mediated-transfers-scenario/raiden-scenario-player-analysis.csv>`_
- JSON-Summary: `<example/mediated-transfers-scenario/raiden-scenario-player-analysis.json>`_

Limitations/Caveats
^^^^^^^^^^^^^^^^^^^

> - the duration of a task is done by looking at the next event in the list of events (sorted by time)

Author: It is still working like that. I just wanted to leave that as a note in case someone occurs a strange duration somewhere. :-/

- parallel task processing might not be visualized correctly

> - the number of ticks on the y-axis can be too much when too many tasks have been emitted in the log. In that case uncomment the ``showticklabels`` line within the ``draw_gantt()`` method.

Reviewer 3: Should this be a cli option?

Author: Can be, but I'd say that this has a ultra low prio right now.
